### PR TITLE
feat: Eraser toggle to switch back to the previous tool

### LINKED ISSUE #4978: toggling the eraser should switch back to the previous tool instead of the select tool
Toggling the eraser should switch back to the previous tool (i.e. the one before clicking eraser) instead of switching to the select tool. You can configure the stylus button to toggle the eraser. If toggling would switch to the original tool (e.g. freedraw) instead of the select tool, that would remove friction from the workflow.

### PR DESCRIPTION
**Pull request by zsviczian**
resolves #4978

### CODE REVIEW

**Comment by dwelle** [comment]
Should we switch to last active tool when hitting `esc` as well? We're kinda breaking new ground here, so not much precedent to go by.

My vote is yes.

File: src/types.ts [hunk 1]
```diff
@@ -77,7 +77,10 @@ export type AppState = {
   // (e.g. text element when typing into the input)
   editingElement: NonDeletedExcalidrawElement | null;
   editingLinearElement: LinearElementEditor | null;
-  activeTool: { type: typeof SHAPES[number]["value"] | "eraser" };
+  activeTool: {
+    type: typeof SHAPES[number]["value"] | "eraser";
+    typeBeforeEraser?: typeof SHAPES[number]["value"] | "eraser";
```
Comment (dwelle) [inline src/types.ts:82]: `typeBeforeEraser` could be confusing. Let's call it `lastActiveToolBeforeEraser`. It doesn't matter the "type" doesn't quite correlate with the naming, as long as the naming is more clear.

Also, the `| "eraser"` should be removed from here.

Comment (zsviczian) [reply src/types.ts:82]: I'll rename the variable.

I also wanted to remove ` | "eraser"`, but then I got a type error here:

https://github.com/excalidraw/excalidraw/blob/db6b335373a9c543304925331f4565acedbaaf90/src/actions/actionCanvas.tsx#L308-L310

**Comment by zsviczian** [comment]
> Should we switch to last active tool when hitting esc as well? We're kinda breaking new ground here, so not much precedent to go by.

My vote is also yes. Will add it later today.

File: src/types.ts (hunk 1 above)
Comment (zsviczian) [reply src/types.ts]: Strangely, when `lastActiveToolBeforeEraser` does not include "eraser" in the type definition I get a type error on the above code when using `isEraserActive(appState)`, but not when I use this expression directly `appState.activeTool.type === "eraser"`

**Comment by dwelle** [comment]
Instead of coupling the `actionFinalize` and `actionEraser` together with those checks, can't we just use the `lastActiveToolBeforeEraser` when updating `activeTool` and `activeTool.type === "eraser"` here?

https://github.com/excalidraw/excalidraw/blob/f242721f3b77a137d8cd04cef1d749463ee71742/src/actions/actionFinalize.tsx#L141-L143

File: src/actions/actionCanvas.tsx [hunk 2]
```diff
@@ -301,12 +301,25 @@ export const actionErase = register({
         ...appState,
         selectedElementIds: {},
         selectedGroupIds: {},
-        activeTool: { type: isEraserActive(appState) ? "selection" : "eraser" },
+        activeTool: {
+          type: isEraserActive(appState)
+            ? appState.activeTool.lastActiveToolBeforeEraser ?? "selection"
+            : "eraser",
+          lastActiveToolBeforeEraser:
+            appState.activeTool.type === "eraser" //node throws incorrect type error when using isEraserActive()
```
Comment (ad1992) [inline src/actions/actionCanvas.tsx:309]:
```suggestion
            appState.activeTool.type === "eraser" 
```
This comment can be removed as its a typescript thingy

**Review by ad1992 — APPROVED** [review]

**Comment by zsviczian** [comment]
shoot… I deployed before noticing the comments :(
both comments are valid, I will create another quick PR to clean it up.
sorry.

**Pull request merged**